src/readable_compacts_transform_helper.js

```diff
@@ -1,117 +1,273 @@
+    }
+    let field = fields.filter(itemFilter(conditions))[0];
+    if (field == undefined) {
+        throw new Error('Message ' + msgName + ' does not has corresponding field with conditions:' + JSON.stringify(conditions));
+    }
+    return {
+        'id': field['id'],
+        'type': field['type'],
+        'rule': field['rule'],
+        'name': field['name'],
+        'keytype':field['keytype'],
+    };
+};
+
+const getQualifiedMsgName = (msgName, pkgName) => {
+    if (msgName.indexOf('.') == -1) {
+        if (pkgName != '') {
+            msgName = pkgName + '.' + msgName;
+        }
+    }
+    return msgName;
+};
+
+const comToLong = (com) =>{
+    return new Long(com, 0 , true);
+};
+
+const hextoString = (hex) => {
+    let bf = ByteBuffer.fromHex(hex);
+    return bf.readString(bf.limit);
+};
+
+const first16ComsToString = (coms, transform)=>{
+    let result ='';
+    for (let ii = 0; ii <16 ; ii++){
+        let component = coms.shift();
+        if (component.toInt() != 0){
+            result = result + transform(component);                       }
+    }
+    return result;
+};
+
+const doCompactsReadableMapping = (messageFieldsMapping, msgName, pkgName, compactComponents) => {
+    if (compactComponents.length == 0){
+        return '';
+    }
+    let components = [].concat(compactComponents);
+    let component = components.shift();
+    let qualifiedMsgName = getQualifiedMsgName(msgName, pkgName);
+    let field = getField({
+        'id': component.toInt()
+    }, qualifiedMsgName, messageFieldsMapping);
+    let readableName = field['name'];
+    let tmp = '';
+    if (field['rule'] === 'repeated') {
+        readableName = readableName + '[' + components.shift().toString() + ']';
+    } else if (field['rule'] === 'map'){
+        switch (field['keytype']){
+        case 'string':
+            tmp = first16ComsToString(components,(com)=>{ return hextoString(com.toString(16));});
+            readableName =  readableName + '[' + tmp + ']'; 
+            break;
+        case 'bytes':
+            tmp = first16ComsToString(components,(com)=>{ return com.toString(16);});
+            readableName = readableName + '[' + '0x' + tmp + ']';
+            break;
+        case 'int64':
+        case 'int32':
+        case 'sint64':
+        case 'sint32':
+        case 'uint64':
+        case 'uint32':
+        case 'fixed64':
+        case 'fixed32':
+            tmp = first16ComsToString(components,(com)=>{ return hextoString(com.toString(16));});
+            readableName = readableName + '[' + tmp + ']';
+            break;  
+        default:
+            throw new Error(`Invalid key type: ${field['keytype']}`);
+        }
+    }
+    let remains = doCompactsReadableMapping(messageFieldsMapping, field['type'], pkgName, components);
+    if (remains !='')  {
+        readableName = readableName + '.' + remains;
+    }   
+    return readableName;
+};
+
+const strToComponents = (str) =>{
+    if (str.length > 128){
```

Comment:
This is not the correct behavior yet. If a string is used as a key in a map, we need to add our own protobuf extension to the field to specify the max length of the key and use that length, not 128 bytes.

Reply:
Why is it 128 and not 64 bytes here?

Reply:
here we have a few choices when golang side do encoding for map key (string or bytes):
1 padding the leading space so whole length is the max length of the key(protobuf extensions)
2 the leading compact component will be a length value (numbers of remaining components), with following components ( if key is string, one component(uint64) can contain 8 ascii char (0x**), if key is bytes, one components can contain 8 bytes), in this way it is impossible to detect leading 0x00 happened in bytes key
3 some subtle encoding  

in the js code, option 1 is used because the encoding rule is not finalized (I guess?)

Reply:
The goal is to go with 1) and zero pad everything. I don't understand why leading zeroes couldn't be added though. The length of the compact key should always be fixed based on the protobuf option set on the field. E.g. a string with length 5 would be 5 bytes long (NB: bytes not characters, so unicode characters taking up more than one byte would so too in the key).

Reply:
option "key_max_length" processing logic has been committed

Reply:
Ok, I understand what you're talking about now. In the literal notation we also always have to extend the value to the max length. I.e. fill it with 0x0 for every character that is missing. So technically a key that is not of the correct length is invalid and doesn't need to be handled. 

@andremarianiello would actually have to make sure this implemented on the go side.

Reply:
@andremarianiello 's  new commit has implemented this kind of logic
Acctually, I' m worry about map<**bytes**, >, if an original bytes key is started with byte 0x00,0x00,
JS cannot distinguish padding 0x00 0x00 0x00.......0x00  from original leading 0x00,0x00 because they are totoally mixed

Reply:
Even for bytes, the literal and compact bytes would always need to be zero padded. A 4 byte array would always be: 0x0, 0x0, 0x0a for example and never just 0x0a

Reply:
to simplify the discussion, assume following definition (**128** happenes at the same way 
map<bytes, string> bytes_map = 1 [
(propfs.max_key_length) = **16**
]

following 0x stand for binary, not hex string
there is an item which is [0x0,0x0,0x12,0x34]-> "blabla.."

if compacts mode on, when golang side do encoding for the proof of  this item, it will pad 12 of 0x0 for the key,   that is  0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,**0x0,0x0,0x12,0x34** (byte slice),then encode those 16 bytes to 2 uint64,  that is **[0x0,0x1234]** (attention, this is uint64 slice, not string slice) which is part of compact property(stand for the key [0x0,0x0,0x12,0x34]).

when JS part do transform from compacts to literal, as far as key is concerned, what it see is just **[0x0,0x1234]**.  It cannot determine the exact key bacause it just know the max length (16, 2 uint64), it does not know the actual length. 

string map does not suffer from this problem, because binary representation of a string will not begin with byte 0x00,0x00...  .    so JS part can remove lead padding 0x00...  safely

Reply:
I'm shortening your example to make it a little bit simpler:

```protobuf,
map<bytes, string> bytes_map = 1 [
(propfs.max_key_length) = 2
]
```

```go,
// NB this is go code, so the native type would be: []byte
var validLengthKey = []byte{0x00, 0x01}
// -> to literal:
var validLiteralKey = "0x0001"
```

The following is invalid and would be extended by the library when generating the tree:
```go,
var input = []byte{0x01}
// If such a value is fed to the go library, the key should be extended to the correct length
var literalKeyBytes = []byte{0x00, 0x01}
```

Also, we're using uint64 for protobuf field numbers (because that's the max value they can have) and for arrays. We don't need to make the mapping keys a multiple of 64 bits. For the compact type, I'd actually suggest we use https://github.com/MikeMcl/bignumber.js/ and not long. The goal is that the library returns a single array of bytes (most likely a bignumber is the best thing to return) not an array of uint64s.

Reply:
@lucasvo Currently, all compact names are represented by some number of uint64s, which are then encoded into bytes. This made sense when all compact names were derived from uint64s, but now with map keys becoming part of the compact name, I think it might not make sense anymore.

Reply:
considering
map<bytes, string> bytes_map = 5 [
(propfs.max_key_length) = 2
] ,
now assume client of precise go library has two items for this map, and it want to produce 2 proofs for these two items
1.  [0x01] -> 'a'  
2.  [0x00,0x01] -> 'a'

After padding proccessing of go library, both items have same property defined in proof.proto 
for compact mode on, it is  array [0x5, 0x1, 0x61]   (here  0x5 is tag number, 0x61 is acsii of 'a' in hex)
for compact mode off, it is string "bytes_map[0x0001].a"
JS part cannot distinguish these two items as far as key ([0x1] or "0x0001") is concerned
actually, go library cannot distinguish these two proofs also. Although other fields in the proof are not same, but it do not know which item the proof is target for, it is a proof for [0x01] -> 'a'  ? or it is a proof for [0x00,0x01] -> 'a'?

Actually, we need another special processing just as **repeated** does (insert a length component in the compacts) 
for map<bytes,..>, consider following encoding rule,
when go library meet a **map**  item and it should produce components for the key in the proof,
it insert a length component (similar to length component of **repeated**),but this component contain two parts, the first 32 bits means numbers of following components which will contribute to this key, the second 32 bits means actual length (number of bytes) of the key.
Now for above two items, compact property will be:
1 [0x5, 0x1000000**01**,0x1,0x61]
2 [0x5, 0x1000000**02**,0x1,0x61]
literal property will be:
1 "bytes_map[0x01].a"
2 "bytes_map[0x0001].a"

in this way, we do not need "max_key_length" extension (of course if you want to limit the length of key, you can introduce this extension, i do not think it is needed)

As far as "Long" is used instead of "bignumber", JS part need parse(here we use protobufjs library) the binary generatey by proto buffer marshal for **proof** message ,  when it meet a uint64(that is the type of compact component) , it will decoded it as Long
 @lucasvo 
@andremarianiello

Reply:
even we change type of compact component from uint64 to byte, we cannot cope with problem mentioned in above comments when map<bytes,..> is met

Reply:
consider encoding of length of  **repeated** field,
if we use byte compact instead of uint64 compact
we have to introduce length of length concept, because length of **repeated** can be 0xaf,  0x01af and 0x0100af

Reply:
I'm sorry if my earlier comment was not clear. The key will always have the specified length. When creating the compact properties, if the precise-proofs library encounters a shorter key it would extend it to the length specified. To take your examples:

```protobuf,
map<bytes, string> bytes_map = 5 [
(propfs.max_key_length) = 2
];
```
> now assume client of precise go library has two items for this map, and it want to produce 2 proofs for these two items
> 
> [0x01] -> 'a'
> [0x00,0x01] -> 'a'

The library here would complain about the first item ([0x01]) being invalid because it has an invalid length and not actually allow you to generate the tree. You wouldn't end up with a tree that has two leaves:
1) `[0x05, 0x01]`
2) `[0x05, 0x00, 0x01]`

Does that make sense?